src/rt/cover.c

```diff
@@ -368,323 +504,571 @@ static cover_file_t *cover_file(const loc_t *loc)
-            "div.advert a:link { text-decoration: none; }\n"
-            "#body { margin: 0px auto; margin-left: 260px;\n"
-            "        font-family: sans-serif; }\n"
-            "#body h1 { font-size: 16pt; text-align: center; }\n"
-            "#body h2 { font-size: 14pt; text-align: center; }\n"
-            "span.hl_miss { background: #ff6060; }\n"
-            "span.hl_hit { background: #60ff60; }\n"
-            "table.stats { border-collapse: collapse; margin-left: auto;\n"
-            "              margin-right: auto; }\n"
-            "table.stats td { padding-right: 10px; }\n"
-            "table.stats th { border-bottom: 1px solid black;\n"
-            "                 padding-right: 10px; font-size: 12pt; }\n"
-            "table.stats td.num { text-align: center; }\n"
-            "span.pct_green { color: #339933; }\n"
-            "span.pct_orange { color: #ff8800; }\n"
-            "span.pct_red { color: #cc0000; }\n");
-    fprintf(fp, "  </style>\n</head>\n<body>\n");
-
-    fprintf(fp, "<div class=\"nav\">\n");
-    fprintf(fp, "<h3>Reports</h3>\n");
-    fprintf(fp, "<ul>\n");
-    fprintf(fp, "  <li><a href=\"index.html\">Index</a></li>\n");
-    fprintf(fp, "</ul>\n");
-    fprintf(fp, "<h3>Files</h3>\n");
-    fprintf(fp, "<ul class=\"nav\">\n");
-    for (cover_file_t *f = files; f != NULL; f = f->next)
-      fprintf(fp, "  <li><a href=\"%s\">%s</a></li>\n",
-              cover_file_url(f), f->name);
-    fprintf(fp, "</ul>\n");
-    fprintf(fp, "</div>\n");
-
-
-    fprintf(fp, "<div id=\"body\">\n");
-}
+   cover_pair_t *pair;
+   int *n;
+   int *alloc;
+
+   if (hits) {
+      pair = chain->hits;
+      n = &(chain->n_hits);
+      alloc = &(chain->alloc_hits);
+   } else {
+      pair = chain->miss;
+      n = &(chain->n_miss);
+      alloc = &(chain->alloc_miss);
+   }
 
-static void cover_html_footer(FILE *fp)
-{
-   fprintf(fp, "</div></body>\n</html>\n");
+   pair[*n].tag = tag;
+   pair[*n].line = line;
+   pair[*n].flags = flags;
+   (*n)++;
+
+   if (*n == *alloc) {
+      *alloc = (*alloc) * 2;
+      pair = xrealloc_array(pair, *alloc, sizeof(cover_pair_t));
+   }
 }
 
-static void cover_report_file(cover_file_t *f, const char *dir)
+static cover_tag_t* cover_report_hierarchy(cover_report_ctx_t *ctx,
+                                           const char *dir)
 {
-   char *buf LOCAL = xasprintf("%s" DIR_SEP "%s", dir, cover_file_url(f));
-   FILE *fp = lib_fopen(lib_work(), buf, "w");
-   if (fp == NULL)
-      fatal("failed to create %s", buf);
+   char *hier LOCAL = xasprintf("%s/%s.html", dir, istr(ctx->start_tag->hier));
+   cover_tag_t *tag = ctx->start_tag;
 
-   cover_html_header(fp, "Coverage report for %s", f->name);
+   // TODO: Handle escaped identifiers in hierarchy path!
+   FILE *f = fopen(hier, "w");
+   if (f == NULL)
+      fatal("Failed to open file: %s\n", hier);
+
+   ctx->ch_stmt.hits = xcalloc(1024 * sizeof(cover_pair_t));
```

Comment:
Use `xcalloc_array` instead of multiplying and calling `calloc` as it handles overflows (I know this won't overflow but it's a good general practice).

Reply:
Done